### PR TITLE
add cors_allow_preflight option to a route's policy

### LINKED ISSUE #78: CORS preflight OPTIONS requests should bypass authentication
**Describe the bug** 
Pomerium is requiring authentication for http `OPTIONS` requests, but the [spec](https://www.w3.org/TR/cors/#cross-origin-request-with-preflight-0) says OPTIONS requests should not include user credentials. These requests should be passed downstream without authentication.

**To Reproduce** Steps to reproduce the behavior:

Browser on `https://bob.corp.domain.com/` makes a CORS request to `https://alice.corp.domain.com/` like this:

```
curl 'https://alice.corp.domain.com/api/route' -X OPTIONS -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0' -H 'Accept: */*' -H 'Accept-Language: en-US,en;q=0.5' --compressed -H 'Access-Control-Request-Method: POST' -H 'Access-Control-Request-Headers: content-type' -H 'Origin: https://bob.corp.domain.com' -H 'DNT: 1' -H 'Connection: keep-alive' -H 'Referer: https://bob.corp.domain.com/some/route' -H 'TE: Trailers'
```

Pomerium will try to redirect to the authentication provider.

**Expected behavior**

Pomerium should not require auth for HTTP `OPTIONS` requests.

**Environment:**

- Pomerium version (retrieve with `pomerium --version`): v0.0.3+41c42f5
- Server Operating System/Architecture/Cloud: ubuntu 18.04, running on EC2 behind an elastic ip

Configuration file(s):

Logs(s):

```text
ERR authenticate: failed to load session error="http: named cookie not present" 
WRN authenticate: authenticate error error="http: named cookie not present" 
```

### PR DESCRIPTION
**Pull request by nitper**
Fixes  #78

Add a route-level setting (`false` by default) to allow unauthenticated OPTIONS requests for CORS. See #78 for more information.

**Checklist**:
- [x] updated docs
- [x] unit tests added
- [x] related issues referenced
- [x] check for required headers in preflight requests
- [x] ready for review

### CODE REVIEW

**Comment by nitper** [comment]
@desimone let me know what you think of this approach before I polish it up and update the docs.

File: proxy/proxy.go [hunk 1]
```diff
@@ -210,7 +211,8 @@ func New(opts *Options) (*Proxy, error) {
 	}
 
 	p := &Proxy{
-		mux: make(map[string]http.Handler),
+		mux:      make(map[string]http.Handler),
```
Comment (desimone) [inline proxy/proxy.go]: knew we'd eventually need something like a wrapper struct that would contain a handler, policy, and any other (future) options... Something like `RouteConfig`. 

But I think this makes sense for now.

File: internal/policy/policy.go [hunk 2]
```diff
@@ -26,6 +26,9 @@ type Policy struct {
 
 	Source      *url.URL
 	Destination *url.URL
+
+	// Allow unauthenticated HTTP OPTIONS requests as per the CORS spec
```
[the remark this answers is not in the document]
Comment (nitper) [reply internal/policy/policy.go]: will do!

File: proxy/handlers.go [hunk 3]
```diff
@@ -227,33 +228,40 @@ func (p *Proxy) OAuthCallback(w http.ResponseWriter, r *http.Request) {
...
 func (p *Proxy) Proxy(w http.ResponseWriter, r *http.Request) {
-	err := p.Authenticate(w, r)
-	// If the authenticate is not successful we proceed to start the OAuth Flow with
-	// OAuthStart. If successful, we proceed to proxy to the configured upstream.
-	if err != nil {
-		switch err {
-		case http.ErrNoCookie, sessions.ErrLifetimeExpired, sessions.ErrInvalidSession:
-			log.FromRequest(r).Debug().Err(err).Msg("proxy: starting auth flow")
-			p.OAuthStart(w, r)
+	pol, foundPolicy := p.policy(r)
```
[the remark this answers is not in the document]
Comment (nitper) [reply proxy/handlers.go]: yep - great idea

File: proxy/proxy.go (hunk 1 above)
Comment (nitper) [reply proxy/proxy.go]: I thought about that as well. I was trying to be minimally intrusive, but I think it makes sense to refactor it now.

**Comment by desimone** [comment]
hi @nitper just wanted to see if you need another review  or any help with this PR. Thanks again.

**Comment by nitper** [comment]
> hi @nitper just wanted to see if you need another review or any help with this PR. Thanks again.

Oops - I just got distracted with other work. Do you have any thoughts on how I [combined Proxy.mux and Proxy.policy into Proxy.routeConfig](https://github.com/pomerium/pomerium/pull/86/commits/10180430338fa33cefdb3b9c48adc509a24b7e89)?

**Comment by desimone** [comment]
@nitper no worries, just wanted to make sure I wasn't blocking you. I think your approach for `routeConfig` looks great.

**Comment by nitper** [comment]
@desimone I'm ready for a final review now - also I think the build needs to be re-run. Looks like it failed pulling down dependencies.

Also let me know if you think I should add an example to `policy.example.yaml` or anywhere else.

File: proxy/handlers.go [hunk 4]
```diff
@@ -223,36 +227,62 @@ func (p *Proxy) OAuthCallback(w http.ResponseWriter, r *http.Request) {
...
+	return false
+}
+
+// isCORSPreflight inspects the request to see if this is a valid CORS preflight request.
+// These checks are not exhaustive, because the proxied server should be verifying it as well.
+//
+// See https://www.html5rocks.com/static/images/cors_server_flowchart.png for more info.
+func isCORSPreflight(r *http.Request) bool {
+	_, ok1 := r.Header["Access-Control-Request-Method"]
+	_, ok2 := r.Header["Origin"]
```
Comment (desimone) [inline proxy/handlers.go]: I wonder if it is safer to use r.Header.Get("X") here? 

See : 
- https://tachingchen.com/blog/pitfall-of-golang-header-operation/
- https://golang.org/pkg/net/http/#Header.Get

File: proxy/handlers.go [hunk 5]
```diff
@@ -223,36 +227,62 @@ func (p *Proxy) OAuthCallback(w http.ResponseWriter, r *http.Request) {
...
+
+	return false
+}
+
+// isCORSPreflight inspects the request to see if this is a valid CORS preflight request.
+// These checks are not exhaustive, because the proxied server should be verifying it as well.
+//
+// See https://www.html5rocks.com/static/images/cors_server_flowchart.png for more info.
+func isCORSPreflight(r *http.Request) bool {
+	_, ok1 := r.Header["Access-Control-Request-Method"]
```
Comment (desimone) [inline proxy/handlers.go]: You mentioned in your initial issue that the OPTIONS request requires three headers: `Access-Control-Request-Method`, `Access-Control-Request-Headers`, and the `Origin header`.

Is `Access-Control-Request-Headers` required?

See:
- https://www.w3.org/TR/cors/#cross-origin-request-with-preflight-0

File: proxy/handlers.go (hunk 4 above)
Comment (nitper) [reply proxy/handlers.go]: Good call - thanks for the links. Fixed.

File: proxy/handlers.go (hunk 5 above)
Comment (nitper) [reply proxy/handlers.go]: I believe `Access-Control-Request-Headers` is not required.

See steps 3 and 4 of https://fetch.spec.whatwg.org/#cors-preflight-fetch, which seem to indicate that if the request does not contain any [`CORS-unsafe request-header names`](https://fetch.spec.whatwg.org/#cors-unsafe-request-header-names), then `Access-Control-Request-Headers` may be omitted. 

[This flowchart](https://www.html5rocks.com/static/images/cors_server_flowchart.png) appears to corroborate the above.

**Comment by nitper** [comment]
Current branch is working in my live setup - here's a sample request generated from the frontend library (URLs obscured):

```bash
curl 'https://destination.corp.example.com/post' -X OPTIONS -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0' -H 'Accept: */*' -H 'Accept-Language: en-US,en;q=0.5' --compressed -H 'Access-Control-Request-Method: POST' -H 'Access-Control-Request-Headers: content-type' -H 'Referer: https://origin.corp.example.com/asdf' -H 'Origin: https://origin.corp.example.com' -H 'DNT: 1' -H 'Connection: keep-alive' -H 'TE: Trailers'
```


Here's my policy:
```yaml
- from: origin.corp.example.com
  to: http://origin/
  allowed_groups:
  - developers@example.com
- from: destination.corp.example.com
  to: http://destination/
  allowed_groups:
  - developers@example.com
  cors_allow_preflight: true
```

the pomerium log shows
```
pomerium        | 1:40PM DBG proxy: skipping authentication for valid CORS preflight request
```

File: proxy/handlers.go (hunk 5 above)
Comment (desimone) [reply proxy/handlers.go]: Great thanks.

**Comment by desimone** [comment]
LGTM! Do you mind squashing the commits then I can merge. Thanks again @nitper !

**Comment by nitper** [comment]
> LGTM! Do you mind squashing the commits then I can merge. Thanks again @nitper !

Done - thanks for the support!

**Pull request merged**